{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## iGEM labeling functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import nltk\n",
    "import regex as re\n",
    "import ast\n",
    "from collections import defaultdict\n",
    "from snorkel.labeling import labeling_function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### setting up abstain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ABSTAIN = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_lfs = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### labeling functions + small tests written"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_solution_words\n",
    "# If any of the words is in a list of solution terms, we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "solution_terms = [\"buffer\", \"diluent\", \"solute\", \"solvent\", \"saturated\", \"unsaturated\", \"saturating\",\n",
    "                    \"saturable\"]\n",
    "@labeling_function()\n",
    "def includes_solution_words(x):\n",
    "    for word in solution_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_solution_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_mixture_words\n",
    "# If any of the words is in a list of mixture terms, we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "mixture_terms = [\"suspended\", \"mixture\", \"heterogenous\", \"homogeneous\"]\n",
    "@labeling_function()\n",
    "def includes_mixture_words(x):\n",
    "    for word in mixture_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_mixture_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_physical_words\n",
    "# If there are common physical terms, we label FALSE\n",
    "physical_terms = [\"detection\", \"composed\", \"weight\", \"characteristic\", \"metal\", \"express\",\n",
    "    \"characterization\", \"color\", \"metalic\", \"consists\", \"pure\", \"compose\", \"assay\", \"mm\", \"bound\",\n",
    "    \"permeable\", \"signal\", \"bind\", \"property\", \"stored\", \"released\", \"capacity\", \"resistance\", \"mol\"]\n",
    "@labeling_function()\n",
    "def includes_physical_words(x):\n",
    "    for word in physical_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_physical_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_genetic_words\n",
    "# If one of the chemicals is in a list of genetic terms such as reductase, dna, gene, we label FALSE\n",
    "genetic_terms = [\"reductase\", \"dna\", \"gene\", \"allele\", \"locus\", \"genotype\", \"phenotype\", \"dominant\", \"recessive\", \"additive\", \"phenoset\",\n",
    "    \"diallelic\", \"multiallelic\", \"polyallelic\", \"monomorphic\", \"monoallelic\", \"polymorphism\", \"complex\", \"trait\", \"multifactorial\",\n",
    "    \"polygenic\", \"monogenic\", \"mixed model\", \"transmission probability\", \"transition probability\", \"epistasis\", \"interaction\", \"pleiotropy\",\n",
    "    \"quantitative trait locus\", \"probit\", \"logit\", \"penetrance\", \"transformation\", \"scale of measurement\", \"identity by descent\", \"identity in state\",\n",
    "    \"haplotype\", \"phase\", \"multilocus\", \"genotype\", \"allelic\", \"association\", \"linkage\", \"disequilibrium\", \"gametic\", \"rna\"]\n",
    "@labeling_function()\n",
    "def includes_genetic_words(x):\n",
    "    for word in genetic_terms:\n",
    "        if (word in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_genetic_words)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_structural_words\n",
    "# If there are common structural terms, we label FALSE\n",
    "structural_terms = [\"loop\", \"sequence\", \"encodes\", \"code\", \"codon\", \"dna\", \"rna\", \"pair\", \"group\", \"active site\",\n",
    "    \"bond\", \"chain\", \"gene\", \"structure\", \"structural\", \"encoding\", \"cdna\", \"cluster\"]\n",
    "@labeling_function()\n",
    "def includes_structural_words(x):\n",
    "    for word in structural_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_structural_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_general_chemical_words\n",
    "# If any of the words is in a list of common chemical terms such as amino acid, sugar, adenosine, amide, adenine, etc..., we label FALSE\n",
    "# https://www.hach.com/chemGlossary   \n",
    "chemical_terms = [\"amino acid\", \"sugar\", \"adenosine\", \"amide\", \"adenine\", \"algaecide\", \"amines\", \"base\", \"biocides\",\n",
    "    \"clarifier\", \"hydrocarbon\", \"molecule\", \"nutrients\", \"polymer\", \"peptide\", \"polypeptide\",\n",
    "    \"tag\", \"functional\", \"electron\", \"cofactor\", \"gas\"]\n",
    "@labeling_function()\n",
    "def includes_general_chemical_words(x):\n",
    "    for word in chemical_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_general_chemical_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_functional_group\n",
    "# If there is a close mention of a functional chemical group, we label FALSE (as it is more likely descriptive of a structure than of a reaction)\n",
    "# https://www.masterorganicchemistry.com/2010/10/06/functional-groups-organic-chemistry/\n",
    "common_functional_groups = [\"alkane\", \"alkene\", \"alkyne\", \"benzene ring\", \"phenyl\", \"amine\", \"alcohol\", \"ether\", \"alkyl halide\", \"thiol\",\n",
    "    \"aldehyde\", \"ketone\", \"ester\", \"carboxylic acid\", \"amide\", \"nitrile\", \"epoxide\", \"disulfide\", \"imine\", \"acid chloride\", \"anhydride\", \"nitro\",\n",
    "    \"sulfide\", \"thioether\", \"group\", \"functional\"]\n",
    "@labeling_function()\n",
    "def includes_functional_group(x):\n",
    "    for group in common_functional_groups:\n",
    "        if (group in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_functional_group)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_amino_acid\n",
    "# If there is an amino acid mentioned, we label FALSE\n",
    "amino_acids = ['val', 'ile', 'leu', 'glu', 'gln', \\\n",
    "    'asp', 'asn' 'his' 'trp', 'phe', 'tyr',    \\\n",
    "    'arg', 'lys', 'ser', 'thr', 'met', 'ala',    \\\n",
    "    'gly', 'pro', 'cys', \"amino\"]\n",
    "@labeling_function()\n",
    "def includes_amino_acid(x):\n",
    "    for group in amino_acids:\n",
    "        if (group in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_amino_acid)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_paper_artifacts\n",
    "# If there are common words from paper headers/footers, we label FALSE\n",
    "common_terms = [\"university\", \"univ\", \"pharma\", \"avenue\", \"street\", \"road\", \"department\", \"usa\", \"reference\", \"ref\",\n",
    "    \"keyword\", \"article\", \"http\", \"png\", \"jpg\", \"journal\", \"(20\", \"(19\"]\n",
    "@labeling_function()\n",
    "def includes_paper_artifacts(x):\n",
    "    for word in common_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_paper_artifacts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_no_terms\n",
    "# If there are words stopping or with no change, we label FALSE\n",
    "common_no_terms = [\"inactivated\", \"unaffected\", \"inactive\", \"inactivates\", \"stops\", \"prevent\", \"inhibit\",\n",
    "    \"denature\", \"block\"]\n",
    "@labeling_function()\n",
    "def includes_no_terms(x):\n",
    "    for word in common_no_terms:\n",
    "        if (word in x[0]):\n",
    "            return (False)\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_no_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def helper_sep_chems_with_or(chemicals):\n",
    "    final = \"\"\n",
    "    for chem in chemicals:\n",
    "        if (final == \"\"):\n",
    "            final += re.escape(chem)\n",
    "        else:\n",
    "            final += \"|\" + re.escape(chem)\n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_adjacent_mentions\n",
    "# If the chemicals are adjacent, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_adjacent_mentions(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \") (\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_adjacent_mentions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_or\n",
    "# If the chemicals are separated by or, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_sep_or(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \") or (\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_or)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_and\n",
    "# If the chemicals are separated by and, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_sep_and(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \") and (\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_and)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_comma\n",
    "# If the chemicals are separated by a comma, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_sep_comma(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \"), (\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_comma)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_sep_via\n",
    "# If the chemicals are separated by via, in, etc, we label FALSE\n",
    "via_terms = \"via|in\"\n",
    "@labeling_function()\n",
    "def structure_sep_via(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \") (\" + via_terms + \") (\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_via)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_sym\n",
    "# If the chemicals are separated by a single character, we label FALSE  \n",
    "@labeling_function()\n",
    "def structure_sep_sym(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + \").(\" + chemicals + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_via)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_followed_by_ase\n",
    "# If one of the chemicals is followed by a word that ends with -ase, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_followed_by_ase(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + r\") \\w*ase\\b\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_followed_by_ase)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_followed_by_noun\n",
    "# If one of the chemicals is followed by a noun, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_followed_by_noun(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    if (chemicals == \"\"):\n",
    "        return ABSTAIN\n",
    "    structure = \"(\" + chemicals + r\") \\w+\"\n",
    "    for match in re.finditer(structure, sentence):\n",
    "        if (nltk.pos_tag([match.group(0).split()[1]])[0][1] == \"NN\"):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_followed_by_noun)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_verb\n",
    "# If the chemicals are separated by a verb, we label TRUE\n",
    "@labeling_function()\n",
    "def structure_sep_verb(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    if (chemicals == \"\"):\n",
    "        return ABSTAIN\n",
    "    structure = \"(\" + chemicals + r\") \\w+ (\" + chemicals + \")\"\n",
    "    for match in re.finditer(structure, sentence):\n",
    "        if (re.match(r\"(\\bVB|NNS)\", nltk.pos_tag([match.group(0).split()[1]])[0][1])):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_verb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_adverb\n",
    "# If the chemicals are separated by a adverb, we label FALSE\n",
    "@labeling_function()\n",
    "def structure_sep_adverb(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    if (chemicals == \"\"):\n",
    "        return ABSTAIN\n",
    "    structure = \"(\" + chemicals + r\") \\w+ (\" + chemicals + \")\"\n",
    "    for match in re.finditer(structure, sentence):\n",
    "        if (re.match(r\"(\\bRB)\", nltk.pos_tag([match.group(0).split()[1]])[0][1])):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_adverb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "structure_sep_adverb([\"cassie crazily night away\", [\"cassie\", \"night\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_oxidation_words\n",
    "# If the sentence contains oxidation words, we label True\n",
    "oxidation_terms = \"oxidiz|oxidis|redox|reduc|rust|corrod|oxygen\"\n",
    "@labeling_function()\n",
    "def includes_oxidation_words(x):\n",
    "    structure = \"(\" + oxidation_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_oxidation_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_combustion_words\n",
    "# If the sentence contains combustion words, we label True\n",
    "combustion_terms = \"combust|burn|explod|gas-form\"\n",
    "@labeling_function()\n",
    "def includes_combustion_words(x):\n",
    "    structure = \"(\" + combustion_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_combustion_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_neutralization_words\n",
    "# If the sentence contains neutralization words, we label True\n",
    "neutralization_terms = \"neutraliz|titrat|buffer|gas-form\"\n",
    "@labeling_function()\n",
    "def includes_neutralization_words(x):\n",
    "    structure = \"(\" + neutralization_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_neutralization_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LF_includes_catalyze_words\n",
    "# If the sentence contains catalyze words, we label True\n",
    "# catalyze_words = [\"catalyze\", \"catalyst\", \"catalyse\", \"catalysing\", \"catalyzing\"]\n",
    "# @labeling_function()\n",
    "# def includes_catalyze_words(x):\n",
    "#     for word in catalyze_words:\n",
    "#         if(word in x[0]):\n",
    "#             return True\n",
    "#     return ABSTAIN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_combination_words\n",
    "# If the sentence contains combination words, we label True\n",
    "combination_terms = \"combin\"\n",
    "@labeling_function()\n",
    "def includes_combination_words(x):\n",
    "    structure = \"(\" + combination_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_combination_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_decomposition_words\n",
    "# If the sentence contains decomposition words, we label True\n",
    "decomposition_terms = \"decompos\"\n",
    "@labeling_function()\n",
    "def includes_decomposition_words(x):\n",
    "    structure = \"(\" + decomposition_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_decomposition_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_replacement_words\n",
    "# If the sentence contains decomposition words, we label True\n",
    "replacement_terms = \"replac\"\n",
    "@labeling_function()\n",
    "def includes_replacement_words(x):\n",
    "    structure = \"(\" + replacement_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_replacement_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_reaction_words\n",
    "# If the sentence contains reactions words, we label True\n",
    "reaction_terms = \"conver|yield|produc|mak|creat|synthesiz|synthesis|transform|ferment|break|displac|exchang\" + \\\n",
    "                 \"|precipit|transfer|through|produc|activat|revers|form|ation|metaboliz|metabolis|generat|hydroly\" + \\\n",
    "                 \"|lead|caus|methyl|result|modif|revers|from\"\n",
    "@labeling_function()\n",
    "def includes_reaction_words(x):\n",
    "    structure = \"(\" + reaction_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_reaction_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_react\n",
    "# If the sentence contains react, we label True\n",
    "@labeling_function()\n",
    "def includes_react(x):\n",
    "    if(\"react\" in x[0]):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_react)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_react_sym\n",
    "# If the sentence contains react, we label True\n",
    "@labeling_function()\n",
    "def includes_react_sym(x):\n",
    "    if(\"-->\" in x[0]):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_react_sym)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_reaction_component_words\n",
    "# If the sentence contains reaction components words, we label True\n",
    "reaction_component_terms = [\"substrate\", \"product\", \"reactant\", \"step\"]\n",
    "@labeling_function()\n",
    "def includes_reaction_component_words(x):\n",
    "    for word in reaction_component_terms:\n",
    "        if(word in x[0]):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_reaction_component_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_comparison_words\n",
    "# If the sentence contains comparison words, we label False\n",
    "comparison_terms = \"similar|more|greater|less|increas|decreas|compar|differ|relativ|better|time|than\"\n",
    "@labeling_function()\n",
    "def includes_comparison_words(x):\n",
    "    structure = \"(\" + comparison_terms + \")\"\n",
    "    if (re.search(structure, x[0])):\n",
    "        return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_comparison_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_concentration\n",
    "# If the sentence contains react, we label True\n",
    "@labeling_function()\n",
    "def includes_concentration(x):\n",
    "    if(\"concentration\" in x[0]):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_concentration)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_measure_words\n",
    "# If the sentence contains measure words, we label False\n",
    "measure_terms = [\"high\", \"low\", \"ph\", \"stability\", \"corelated\", \"more\", \"less\", \"level\", \"degree\", \"time\",\n",
    "                \"measure\"]\n",
    "@labeling_function()\n",
    "def includes_measure_words(x):\n",
    "    for word in measure_terms:\n",
    "        if(word in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_measure_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_experiment_words\n",
    "# If the sentence contains experiement words, we label False\n",
    "experiment_terms = [\"mice\", \"cell\", \"mouse\", \"ovary\", \"male\", \"female\", \"animal\", \"study\", \"method\",\n",
    "    \"test\", \"treat\", \"protection\", \"brain\", \"nerve\", \"human\", \"tissue\", \"fetal\", \"vitro\", \"studies\",\n",
    "    \"membrane\", \"strain\", \"mutant\", \"regulate\", \"dependent\", \"drug\", \"therapy\", \"oral\", \"test\", \"autoantigen\"]\n",
    "@labeling_function()\n",
    "def includes_experiment_words(x):\n",
    "    for word in experiment_terms:\n",
    "        if(word in x[0]):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_experiment_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_next_to_conversion_words\n",
    "# If the sentence contains to, from, into, etc., we label True\n",
    "conversion_terms_next = \"to|from|into|becom|became|by\"\n",
    "@labeling_function()\n",
    "def structure_next_to_conversion_words(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"((\" + chemicals + r\") (\" + conversion_terms_next + r\")\\b|((\" + conversion_terms_next + r\")\\b (\" + chemicals + \")))\"\n",
    "    if (re.search(structure, sentence)):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_next_to_conversion_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_sep_conversion_words\n",
    "# If the sentence contains to, from, into, etc., we label True\n",
    "conversion_terms_sep = \"to|into\"\n",
    "@labeling_function()\n",
    "def structure_sep_conversion_words(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + chemicals + r\") (\" + conversion_terms_sep + r\") (\" + chemicals + \")\"\n",
    "    if (re.search(structure, sentence)):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_sep_conversion_words)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_conversion_by\n",
    "# If the sentence contains to, from, into, etc., we label True\n",
    "conversion_terms_by = \"to|from\"\n",
    "@labeling_function()\n",
    "def structure_conversion_by(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"(\" + conversion_terms_by + \") (\" + chemicals + \") by\"\n",
    "    if (re.search(structure, sentence)):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_conversion_by)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure_conversion_of\n",
    "# If the sentence contains to, from, into, etc., we label True\n",
    "conversion_terms_of = \"ation of|sion of|ism of\"\n",
    "@labeling_function()\n",
    "def structure_conversion_of(x):\n",
    "    sentence = x[0].replace(',', '')\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure = \"((\" + conversion_terms_of + \") (\" + chemicals + \")|of (\" + chemicals + \") (\" + conversion_terms_sep + \"))\"\n",
    "    if (re.search(structure, sentence)):\n",
    "            return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_conversion_of)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_one_chem\n",
    "# If the sentence contains one identified chemical, we label False\n",
    "@labeling_function()\n",
    "def includes_one_chem(x):\n",
    "    if(len(x[1]) == 1):\n",
    "            return False\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_one_chem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # includes_more_than_one_chem\n",
    "# # If the sentence contains more than one identified chemical, we label True\n",
    "# @labeling_function()\n",
    "# def includes_more_than_one_chem(x):\n",
    "#     if(len(x[1]) > 1):\n",
    "#             return True\n",
    "#     return ABSTAIN\n",
    "\n",
    "# all_lfs.append(includes_more_than_one_chem)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### max's labeling functions (updated 11/30/21) ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# includes_interconvert_sym\n",
    "# If the sentence contains the interconvert symbol, we label TRUE\n",
    "#\n",
    "# POTENTIAL CHANGE: symbol must be inbetween two chemicals?\n",
    "@labeling_function()\n",
    "def includes_interconvert_sym(x):\n",
    "    if(\"<->\" in x[0]):\n",
    "        return True\n",
    "    if(\"<-->\" in x[0]):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(includes_interconvert_sym)\n",
    "\n",
    "\n",
    "# structure_jtsui_pattern_1\n",
    "# If part of the sentence contains the specific structure\n",
    "# [trigger1] <0,1> chemical <0,1> [transition] chemical, we label True\n",
    "TRANS = \"from|to|into|by|are|yield\"\n",
    "TRIG1 = \"phosphoryl|condens|hydrolys|metabol|reduc|conver|produc|form|oxid|transform|bioconver|synthes|react|interconver\"\n",
    "@labeling_function()\n",
    "def structure_jtsui_pattern_1(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure1 = \"\\b\\w*\" + TRIG1 + \"\\w*\\b (\" + chemicals + \") \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure2 = \"\\b\\w*\" + TRIG1 + \"\\w*\\b \\w+ (\" + chemicals + \") \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure3 = \"\\b\\w*\" + TRIG1 + \"\\w*\\b (\" + chemicals + \") \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure4 = \"\\b\\w*\" + TRIG1 + \"\\w*\\b \\w+ (\" + chemicals + \") \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    if (re.search(structure1, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure2, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure3, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure4, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_jtsui_pattern_1)\n",
    "\n",
    "# structure_jtsui_pattern_2\n",
    "# If part of the sentence contains the specific structure\n",
    "# chemical <0,1> [trigger2] <0,1> [transition] <0,1> chemical, we label True\n",
    "TRIG2 = 'conver|oxid|produc|interconver'\n",
    "@labeling_function()\n",
    "def structure_jtsui_pattern_2(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure1 = \"(\" + chemicals + \")\" \"\\b\\w*\" + TRIG2 + \"\\w*\\b \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure2 = \"(\" + chemicals + \") \\w+ \\b\\w*\" + TRIG2 + \"\\w*\\b \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure3 = \"(\" + chemicals + \") \\w+ \\b\\w*\" + TRIG2 + \"\\w*\\b \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure4 = \"(\" + chemicals + \") \\w+ \\b\\w*\" + TRIG2 + \"\\w*\\b \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b \\w+ (\" + chemicals + \")\"\n",
    "    structure5 = \"(\" + chemicals + \") \\b\\w*\" + TRIG2 + \"\\w*\\b \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure6 = \"(\" + chemicals + \") \\b\\w*\" + TRIG2 + \"\\w*\\b \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b \\w+ (\" + chemicals + \")\"\n",
    "    structure7 = \"(\" + chemicals + \") \\b\\w*\" + TRIG2 + \"\\w*\\b \\b\\w*[\" + TRANS + \"]\\w+\\b \\w+ (\" + chemicals + \")\"\n",
    "    structure8 = \"(\" + chemicals + \") \\w+ \\b\\w*\" + TRIG2 + \"\\w*\\b \\b\\w*[\" + TRANS + \"]\\w+\\b \\w+ (\" + chemicals + \")\"\n",
    "    if (re.search(structure1, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure2, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure3, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure4, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure5, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure6, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure7, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure8, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_jtsui_pattern_2)\n",
    "\n",
    "# structure_jtsui_pattern_3\n",
    "# If part of the sentence contains the specific structure\n",
    "# chemical [trigger3] <0,1> chemical, we label True\n",
    "TRIG3 = 'yield'\n",
    "@labeling_function()\n",
    "def structure_jtsui_pattern_3(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure1 = \"(\" + chemicals + \") \\b\\w*\" + TRIG3 + \"\\w*\\b \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    structure2 = \"(\" + chemicals + \") \\b\\w*\" + TRIG3 + \"\\w*\\b \\w+ \\b\\w*[\" + TRANS + \"]\\w+\\b (\" + chemicals + \")\"\n",
    "    if (re.search(structure1, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure2, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_jtsui_pattern_3)\n",
    "\n",
    "# structure_jtsui_pattern_4\n",
    "# If part of the sentence contains the specific structure\n",
    "# [trigger4] <0,1> chemical, we label True\n",
    "TRIG4 = 'conver|interconver'\n",
    "@labeling_function()\n",
    "def structure_jtsui_pattern_4(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure1 = \"\\b\\w*\" + TRIG4 + \"\\w*\\b (\" + chemicals + \")\"\n",
    "    structure2 = \"\\b\\w*\" + TRIG4 + \"\\w*\\b of (\" + chemicals + \")\"\n",
    "    if (re.search(structure1, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure2, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_jtsui_pattern_4)\n",
    "\n",
    "# structure_jtsui_pattern_5\n",
    "# If part of the sentence contains the specific structure\n",
    "# chemical [transition5] <0,1> [trigger5] <0,1> chemical, we label True\n",
    "TRIG5 = 'produc|metaboli'\n",
    "TRANS5 = 'is|are'\n",
    "@labeling_function()\n",
    "def structure_jtsui_pattern_5(x):\n",
    "    chemicals = helper_sep_chems_with_or(x[1])\n",
    "    structure1 = \"(\" + chemicals + \") \\b\\w*[\" + TRANS5 + \"]\\w+\\b \\b\\w*\" + TRIG5 + \"\\w*\\b (\" + chemicals + \")\"\n",
    "    structure2 = \"(\" + chemicals + \") \\b\\w*[\" + TRANS5 + \"]\\w+\\b \\w+ \\b\\w*\" + TRIG5 + \"\\w*\\b (\" + chemicals + \")\"\n",
    "    structure3 = \"(\" + chemicals + \") \\b\\w*[\" + TRANS5 + \"]\\w+\\b \\b\\w*\" + TRIG5 + \"\\w*\\b \\w+ (\" + chemicals + \")\"\n",
    "    structure4 = \"(\" + chemicals + \") \\b\\w*[\" + TRANS5 + \"]\\w+\\b \\w+ \\b\\w*\" + TRIG5 + \"\\w*\\b \\w+ (\" + chemicals + \")\"\n",
    "    if (re.search(structure1, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure2, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure3, x[0])):\n",
    "        return True\n",
    "    if (re.search(structure4, x[0])):\n",
    "        return True\n",
    "    return ABSTAIN\n",
    "\n",
    "all_lfs.append(structure_jtsui_pattern_5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### data cleaning stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "greek_alphabet = {\n",
    "    u'\\u0393': 'Gamma',\n",
    "    u'\\u0394': 'Delta',\n",
    "    u'\\u0398': 'Theta',\n",
    "    u'\\u039B': 'Lamda',\n",
    "    u'\\u039E': 'Xi',\n",
    "    u'\\u03A0': 'Pi',\n",
    "    u'\\u03A3': 'Sigma',\n",
    "    u'\\u03A6': 'Phi',\n",
    "    u'\\u03A7': 'Chi',\n",
    "    u'\\u03A8': 'Psi',\n",
    "    u'\\u03A9': 'Omega',\n",
    "    u'\\u03B1': 'alpha',\n",
    "    u'\\u03B2': 'beta',\n",
    "    u'\\u03B3': 'gamma',\n",
    "    u'\\u03B4': 'delta',\n",
    "    u'\\u03B5': 'epsilon',\n",
    "    u'\\u03B6': 'zeta',\n",
    "    u'\\u03B7': 'eta',\n",
    "    u'\\u03B8': 'theta',\n",
    "    u'\\u03B9': 'iota',\n",
    "    u'\\u03BA': 'kappa',\n",
    "    u'\\u03BB': 'lamda',\n",
    "    u'\\u03BC': 'mu',\n",
    "    u'\\u03BD': 'nu',\n",
    "    u'\\u03BE': 'xi',\n",
    "    u'\\u03BF': 'omicron',\n",
    "    u'\\u03C0': 'pi',\n",
    "    u'\\u03C1': 'rho',\n",
    "    u'\\u03C3': 'sigma',\n",
    "    u'\\u03C4': 'tau',\n",
    "    u'\\u03C5': 'upsilon',\n",
    "    u'\\u03C6': 'phi',\n",
    "    u'\\u03C7': 'chi',\n",
    "    u'\\u03C8': 'psi',\n",
    "    u'\\u03C9': 'omega'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_greek(words):\n",
    "    for letter in greek_alphabet.keys():\n",
    "        words = words.replace(letter, greek_alphabet[letter])\n",
    "    return words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chem_into_array(chemicals):\n",
    "    if (chemicals == \"0\"):\n",
    "        return []\n",
    "    chemicals = remove_greek(chemicals)\n",
    "    chemicals = chemicals.lower()\n",
    "    sample_chems = chemicals.replace(\"%20\", \" \")\n",
    "    sample_chems_list = sample_chems.split(\", \")\n",
    "    for index in range(len(sample_chems_list)):\n",
    "        while(sample_chems_list[index][-1] == \",\"):\n",
    "            sample_chems_list[index] = sample_chems_list[index][:-1]\n",
    "    sample_chems_list = list(set(sample_chems_list))\n",
    "    return sample_chems_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sentence_cleaned(sentence):\n",
    "    sentence = remove_greek(sentence)\n",
    "    sentence = sentence.lower()\n",
    "    sentence = sentence.replace(\".\", \"\")\n",
    "    sentence = sentence.replace(\")\", \"\")\n",
    "    sentence = sentence.replace(\"(\", \"\")\n",
    "    sentence = sentence.replace(\";\", \"\")\n",
    "    sentence = sentence.replace(\":\", \"\")\n",
    "    return sentence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### importing data in and cleaning it (this data has truth values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# have this csv file in the same folder\n",
    "sentence_df = pd.read_csv(\"sentence_data_cleaned_csv_fixed.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>truth</th>\n",
       "      <th>substrates</th>\n",
       "      <th>products</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The enzyme cyclo-oxygenase catalyses the oxyge...</td>\n",
       "      <td>prostaglandins</td>\n",
       "      <td>1</td>\n",
       "      <td>arachidonic, acid</td>\n",
       "      <td>prostaglandins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Recently two forms of cyclo-oxygenase have bee...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Constitutive and inducible forms of human cycl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCOX-1 had a specific activity of 18.8 mumol o...</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>of 1500 nmol of O2/nmol of enzyme, whereas hCO...</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sentence       chemicals  truth  \\\n",
       "0  The enzyme cyclo-oxygenase catalyses the oxyge...  prostaglandins      1   \n",
       "1  Recently two forms of cyclo-oxygenase have bee...             NaN      0   \n",
       "2  Constitutive and inducible forms of human cycl...             NaN      0   \n",
       "3  hCOX-1 had a specific activity of 18.8 mumol o...    arachidonate      0   \n",
       "4  of 1500 nmol of O2/nmol of enzyme, whereas hCO...    arachidonate      0   \n",
       "\n",
       "          substrates        products  \n",
       "0  arachidonic, acid  prostaglandins  \n",
       "1                NaN             NaN  \n",
       "2                NaN             NaN  \n",
       "3       arachidonate             NaN  \n",
       "4       arachidonate             NaN  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df = sentence_df[[\"sentence\", \"chemicals\", \"truth\", \"substrates\", \"products\"]]\n",
    "sentence_chem_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/1323417438.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].replace(np.NaN, \"0\")\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].replace(np.NaN, \"0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/436195421.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    1354\n",
       "1      60\n",
       "Name: truth, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df[\"sentence\"] = sentence_chem_df[\"sentence\"].apply(sentence_cleaned)\n",
    "sentence_chem_df[\"truth\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/741021427.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)\n"
     ]
    }
   ],
   "source": [
    "sentence_chem_df[\"chemicals\"] = sentence_chem_df[\"chemicals\"].apply(chem_into_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/4259470197.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sentence_chem_df[\"text\"] = sentence_chem_df[[\"sentence\", \"chemicals\"]].values.tolist()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>truth</th>\n",
       "      <th>substrates</th>\n",
       "      <th>products</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>the enzyme cyclo-oxygenase catalyses the oxyge...</td>\n",
       "      <td>[prostaglandins]</td>\n",
       "      <td>1</td>\n",
       "      <td>arachidonic, acid</td>\n",
       "      <td>prostaglandins</td>\n",
       "      <td>[the enzyme cyclo-oxygenase catalyses the oxyg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>recently two forms of cyclo-oxygenase have bee...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[recently two forms of cyclo-oxygenase have be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>constitutive and inducible forms of human cycl...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[constitutive and inducible forms of human cyc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hcox-1 had a specific activity of 188 mumol of...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[hcox-1 had a specific activity of 188 mumol o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>of 1500 nmol of o2/nmol of enzyme, whereas hco...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[of 1500 nmol of o2/nmol of enzyme, whereas hc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>water uptake of the polymer was only 28 and 02...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[water uptake of the polymer was only 28 and 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1410</th>\n",
       "      <td>degradation of less hydrophilic pec41 with hig...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[degradation of less hydrophilic pec41 with hi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1411</th>\n",
       "      <td>by this mechanism, ce-responsive drug in vitro...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[by this mechanism, ce-responsive drug in vitr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>as expected, less bovine serum albumin bsa was...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[as expected, less bovine serum albumin bsa wa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>in conclusion, this work enabled the in vitro ...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[in conclusion, this work enabled the in vitro...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1414 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence         chemicals  \\\n",
       "0     the enzyme cyclo-oxygenase catalyses the oxyge...  [prostaglandins]   \n",
       "1     recently two forms of cyclo-oxygenase have bee...                []   \n",
       "2     constitutive and inducible forms of human cycl...                []   \n",
       "3     hcox-1 had a specific activity of 188 mumol of...    [arachidonate]   \n",
       "4     of 1500 nmol of o2/nmol of enzyme, whereas hco...    [arachidonate]   \n",
       "...                                                 ...               ...   \n",
       "1409  water uptake of the polymer was only 28 and 02...             [pec]   \n",
       "1410  degradation of less hydrophilic pec41 with hig...                []   \n",
       "1411  by this mechanism, ce-responsive drug in vitro...             [pec]   \n",
       "1412  as expected, less bovine serum albumin bsa was...                []   \n",
       "1413  in conclusion, this work enabled the in vitro ...             [pec]   \n",
       "\n",
       "      truth         substrates        products  \\\n",
       "0         1  arachidonic, acid  prostaglandins   \n",
       "1         0                NaN             NaN   \n",
       "2         0                NaN             NaN   \n",
       "3         0       arachidonate             NaN   \n",
       "4         0       arachidonate             NaN   \n",
       "...     ...                ...             ...   \n",
       "1409      0                NaN             NaN   \n",
       "1410      0                NaN             NaN   \n",
       "1411      0                NaN             NaN   \n",
       "1412      0                NaN             NaN   \n",
       "1413      0                NaN             NaN   \n",
       "\n",
       "                                                   text  \n",
       "0     [the enzyme cyclo-oxygenase catalyses the oxyg...  \n",
       "1     [recently two forms of cyclo-oxygenase have be...  \n",
       "2     [constitutive and inducible forms of human cyc...  \n",
       "3     [hcox-1 had a specific activity of 188 mumol o...  \n",
       "4     [of 1500 nmol of o2/nmol of enzyme, whereas hc...  \n",
       "...                                                 ...  \n",
       "1409  [water uptake of the polymer was only 28 and 0...  \n",
       "1410  [degradation of less hydrophilic pec41 with hi...  \n",
       "1411  [by this mechanism, ce-responsive drug in vitr...  \n",
       "1412  [as expected, less bovine serum albumin bsa wa...  \n",
       "1413  [in conclusion, this work enabled the in vitro...  \n",
       "\n",
       "[1414 rows x 6 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df[\"text\"] = sentence_chem_df[[\"sentence\", \"chemicals\"]].values.tolist()\n",
    "sentence_chem_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### data with no truth values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# have this data one folder up\n",
    "uncleaned_df = pd.read_csv(\"../sentence_annotations_elsevier_pmid_split6.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "uncleaned_no_na_df = uncleaned_df.dropna()\n",
    "uncleaned_several_chem_df = uncleaned_no_na_df.loc[uncleaned_no_na_df[\"chemical_names\"].str.contains(\",\")]\n",
    "uncleaned_several_chem_df = uncleaned_several_chem_df[[\"sentence\", \"chemical_names\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "uncleaned_several_chem_df[\"sentence\"] = uncleaned_several_chem_df[\"sentence\"].apply(sentence_cleaned)\n",
    "uncleaned_several_chem_df[\"chemical_names\"] = uncleaned_several_chem_df[\"chemical_names\"].apply(chem_into_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to get rid of this line if this data should be used\n",
    "re_check_several_chem_uncleaned_df = uncleaned_several_chem_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "re_check_several_chem_uncleaned_df[\"text\"] = re_check_several_chem_uncleaned_df[[\"sentence\", \"chemical_names\"]].values.tolist()\n",
    "brenda_clean_df = re_check_several_chem_uncleaned_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### snorkel code for model (running on data with truth values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.labeling import PandasLFApplier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how to find the location\n",
    "# sentence_chem_df[sentence_chem_df[\"sentence\"] == \"even though the activities of mat and gnmt were elevated, the concentration of liver s-adenosylmethionine was decreased 24%, p<0001 and s-adenosylhomocysteine increased 113%, p<0001 in the dwarf mice\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>truth</th>\n",
       "      <th>substrates</th>\n",
       "      <th>products</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>the enzyme cyclo-oxygenase catalyses the oxyge...</td>\n",
       "      <td>[prostaglandins]</td>\n",
       "      <td>1</td>\n",
       "      <td>arachidonic, acid</td>\n",
       "      <td>prostaglandins</td>\n",
       "      <td>[the enzyme cyclo-oxygenase catalyses the oxyg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>recently two forms of cyclo-oxygenase have bee...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[recently two forms of cyclo-oxygenase have be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>constitutive and inducible forms of human cycl...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[constitutive and inducible forms of human cyc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hcox-1 had a specific activity of 188 mumol of...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[hcox-1 had a specific activity of 188 mumol o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>of 1500 nmol of o2/nmol of enzyme, whereas hco...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[of 1500 nmol of o2/nmol of enzyme, whereas hc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sentence         chemicals  truth  \\\n",
       "0  the enzyme cyclo-oxygenase catalyses the oxyge...  [prostaglandins]      1   \n",
       "1  recently two forms of cyclo-oxygenase have bee...                []      0   \n",
       "2  constitutive and inducible forms of human cycl...                []      0   \n",
       "3  hcox-1 had a specific activity of 188 mumol of...    [arachidonate]      0   \n",
       "4  of 1500 nmol of o2/nmol of enzyme, whereas hco...    [arachidonate]      0   \n",
       "\n",
       "          substrates        products  \\\n",
       "0  arachidonic, acid  prostaglandins   \n",
       "1                NaN             NaN   \n",
       "2                NaN             NaN   \n",
       "3       arachidonate             NaN   \n",
       "4       arachidonate             NaN   \n",
       "\n",
       "                                                text  \n",
       "0  [the enzyme cyclo-oxygenase catalyses the oxyg...  \n",
       "1  [recently two forms of cyclo-oxygenase have be...  \n",
       "2  [constitutive and inducible forms of human cyc...  \n",
       "3  [hcox-1 had a specific activity of 188 mumol o...  \n",
       "4  [of 1500 nmol of o2/nmol of enzyme, whereas hc...  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence_chem_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemical_names</th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>we selected diltiazem, cyclosporin a, and dexa...</td>\n",
       "      <td>[dexamethasone, calcein-am, diltiazem, cyclosp...</td>\n",
       "      <td>[we selected diltiazem, cyclosporin a, and dex...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>croteau r washington state univ, inst biol che...</td>\n",
       "      <td>[geranyl pyrophosphate, limonene, sodium dodec...</td>\n",
       "      <td>[croteau r washington state univ, inst biol ch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>storage tissue h+/substrate stoichiometries fo...</td>\n",
       "      <td>[h+]</td>\n",
       "      <td>[storage tissue h+/substrate stoichiometries f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>briskin dp univ illinois, dept agron, 1201 w g...</td>\n",
       "      <td>[h+]</td>\n",
       "      <td>[briskin dp univ illinois, dept agron, 1201 w ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>from these results and the estimated level of ...</td>\n",
       "      <td>[h+]</td>\n",
       "      <td>[from these results and the estimated level of...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              sentence  \\\n",
       "150  we selected diltiazem, cyclosporin a, and dexa...   \n",
       "271  croteau r washington state univ, inst biol che...   \n",
       "280  storage tissue h+/substrate stoichiometries fo...   \n",
       "281  briskin dp univ illinois, dept agron, 1201 w g...   \n",
       "286  from these results and the estimated level of ...   \n",
       "\n",
       "                                        chemical_names  \\\n",
       "150  [dexamethasone, calcein-am, diltiazem, cyclosp...   \n",
       "271  [geranyl pyrophosphate, limonene, sodium dodec...   \n",
       "280                                               [h+]   \n",
       "281                                               [h+]   \n",
       "286                                               [h+]   \n",
       "\n",
       "                                                  text  \n",
       "150  [we selected diltiazem, cyclosporin a, and dex...  \n",
       "271  [croteau r washington state univ, inst biol ch...  \n",
       "280  [storage tissue h+/substrate stoichiometries f...  \n",
       "281  [briskin dp univ illinois, dept agron, 1201 w ...  \n",
       "286  [from these results and the estimated level of...  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "brenda_clean_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 50212/50212 [12:30<00:00, 66.93it/s]\n"
     ]
    }
   ],
   "source": [
    "# Define the set of labeling functions (LFs)\n",
    "# currently excluding amino_acid and followed_ase and followed_by_noun\n",
    "lfs = [includes_solution_words, includes_mixture_words, includes_physical_words, includes_genetic_words, includes_structural_words,\n",
    "      includes_general_chemical_words, includes_functional_group, includes_paper_artifacts, includes_no_terms, structure_adjacent_mentions,\n",
    "      structure_sep_or, structure_sep_comma, structure_sep_via, structure_sep_sym, structure_sep_adverb, includes_oxidation_words,\n",
    "      structure_sep_verb, structure_sep_conversion_words, includes_combustion_words, includes_neutralization_words, includes_combination_words,\n",
    "      includes_decomposition_words, includes_replacement_words, includes_reaction_words, includes_reaction_component_words, includes_comparison_words,\n",
    "      includes_one_chem, includes_react, includes_measure_words, includes_experiment_words, includes_concentration, structure_sep_and,\n",
    "      structure_next_to_conversion_words, structure_conversion_by, structure_conversion_of, includes_react_sym,\n",
    "      includes_interconvert_sym, structure_jtsui_pattern_1, structure_jtsui_pattern_2,\n",
    "      structure_jtsui_pattern_3, structure_jtsui_pattern_4, structure_jtsui_pattern_5]\n",
    "      # includes_amino_acid, structure_followed_by_ase, structure_followed_by_noun]\n",
    "# from max\n",
    "# includes_interconvert_sym, structure_jtsui_pattern_1, structure_jtsui_pattern_2, structure_jtsui_pattern_3, structure_jtsui_pattern_4, structure_jtsui_pattern_5\n",
    "# removing physical_words increases recall but causes large drop in precision\n",
    "# sep_conversion_word and sep_verb removal increase precision to 0.71 with recall at 0.38\n",
    "# Apply the LFs to the unlabeled training data\n",
    "df_train = brenda_clean_df\n",
    "applier = PandasLFApplier(lfs=lfs)\n",
    "L_train = applier.apply(df=df_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1, -1, -1,  0,  0, -1, -1,  0,  0, -1, -1, -1, -1, -1, -1, -1, -1,\n",
       "       -1, -1, -1, -1, -1, -1,  1, -1, -1, -1, -1,  0, -1, -1, -1,  1, -1,\n",
       "        1, -1, -1,  1, -1, -1, -1, -1])"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L_train[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>includes_solution_words</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>0.053155</td>\n",
       "      <td>0.051940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_mixture_words</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.026826</td>\n",
       "      <td>0.026806</td>\n",
       "      <td>0.024815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_physical_words</th>\n",
       "      <td>2</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.456126</td>\n",
       "      <td>0.452920</td>\n",
       "      <td>0.397734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_genetic_words</th>\n",
       "      <td>3</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.281865</td>\n",
       "      <td>0.281526</td>\n",
       "      <td>0.250856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_structural_words</th>\n",
       "      <td>4</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.312834</td>\n",
       "      <td>0.311519</td>\n",
       "      <td>0.270055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_general_chemical_words</th>\n",
       "      <td>5</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.235342</td>\n",
       "      <td>0.234287</td>\n",
       "      <td>0.204871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_functional_group</th>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.250199</td>\n",
       "      <td>0.249562</td>\n",
       "      <td>0.220505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_paper_artifacts</th>\n",
       "      <td>7</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.132717</td>\n",
       "      <td>0.132419</td>\n",
       "      <td>0.122461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_no_terms</th>\n",
       "      <td>8</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.131980</td>\n",
       "      <td>0.130467</td>\n",
       "      <td>0.113957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_adjacent_mentions</th>\n",
       "      <td>9</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.092906</td>\n",
       "      <td>0.092906</td>\n",
       "      <td>0.082430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_or</th>\n",
       "      <td>10</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.015873</td>\n",
       "      <td>0.015733</td>\n",
       "      <td>0.013602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_comma</th>\n",
       "      <td>11</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.048116</td>\n",
       "      <td>0.047897</td>\n",
       "      <td>0.041385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_via</th>\n",
       "      <td>12</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.003664</td>\n",
       "      <td>0.003605</td>\n",
       "      <td>0.003226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_sym</th>\n",
       "      <td>13</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.174938</td>\n",
       "      <td>0.173664</td>\n",
       "      <td>0.148929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_adverb</th>\n",
       "      <td>14</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.000637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_oxidation_words</th>\n",
       "      <td>15</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.086673</td>\n",
       "      <td>0.086653</td>\n",
       "      <td>0.084522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_verb</th>\n",
       "      <td>16</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.009858</td>\n",
       "      <td>0.009838</td>\n",
       "      <td>0.009619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_conversion_words</th>\n",
       "      <td>17</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.024636</td>\n",
       "      <td>0.024636</td>\n",
       "      <td>0.022783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combustion_words</th>\n",
       "      <td>18</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.001215</td>\n",
       "      <td>0.001215</td>\n",
       "      <td>0.001215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_neutralization_words</th>\n",
       "      <td>19</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.045547</td>\n",
       "      <td>0.045547</td>\n",
       "      <td>0.045547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combination_words</th>\n",
       "      <td>20</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.024855</td>\n",
       "      <td>0.024835</td>\n",
       "      <td>0.024496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_decomposition_words</th>\n",
       "      <td>21</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.001135</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>0.001115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_replacement_words</th>\n",
       "      <td>22</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.008763</td>\n",
       "      <td>0.008703</td>\n",
       "      <td>0.008604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_words</th>\n",
       "      <td>23</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.750737</td>\n",
       "      <td>0.745081</td>\n",
       "      <td>0.727256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_component_words</th>\n",
       "      <td>24</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.146459</td>\n",
       "      <td>0.146061</td>\n",
       "      <td>0.142038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_comparison_words</th>\n",
       "      <td>25</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.236975</td>\n",
       "      <td>0.235601</td>\n",
       "      <td>0.204055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_one_chem</th>\n",
       "      <td>26</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.109297</td>\n",
       "      <td>0.108042</td>\n",
       "      <td>0.090337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_react</th>\n",
       "      <td>27</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.098502</td>\n",
       "      <td>0.098502</td>\n",
       "      <td>0.095256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_measure_words</th>\n",
       "      <td>28</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.576257</td>\n",
       "      <td>0.571298</td>\n",
       "      <td>0.498486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_experiment_words</th>\n",
       "      <td>29</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.365550</td>\n",
       "      <td>0.362981</td>\n",
       "      <td>0.318071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_concentration</th>\n",
       "      <td>30</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.052358</td>\n",
       "      <td>0.052358</td>\n",
       "      <td>0.050944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_and</th>\n",
       "      <td>31</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.086513</td>\n",
       "      <td>0.086035</td>\n",
       "      <td>0.076097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_next_to_conversion_words</th>\n",
       "      <td>32</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.206883</td>\n",
       "      <td>0.206026</td>\n",
       "      <td>0.197443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_conversion_by</th>\n",
       "      <td>33</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.008165</td>\n",
       "      <td>0.008165</td>\n",
       "      <td>0.007190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_conversion_of</th>\n",
       "      <td>34</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.114933</td>\n",
       "      <td>0.114933</td>\n",
       "      <td>0.110511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_react_sym</th>\n",
       "      <td>35</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.000080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_interconvert_sym</th>\n",
       "      <td>36</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_1</th>\n",
       "      <td>37</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.447045</td>\n",
       "      <td>0.446965</td>\n",
       "      <td>0.432347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_2</th>\n",
       "      <td>38</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.152215</td>\n",
       "      <td>0.152215</td>\n",
       "      <td>0.146539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_3</th>\n",
       "      <td>39</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_4</th>\n",
       "      <td>40</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_5</th>\n",
       "      <td>41</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     j Polarity  Coverage  Overlaps  Conflicts\n",
       "includes_solution_words              0      [0]  0.053254  0.053155   0.051940\n",
       "includes_mixture_words               1      [0]  0.026826  0.026806   0.024815\n",
       "includes_physical_words              2      [0]  0.456126  0.452920   0.397734\n",
       "includes_genetic_words               3      [0]  0.281865  0.281526   0.250856\n",
       "includes_structural_words            4      [0]  0.312834  0.311519   0.270055\n",
       "includes_general_chemical_words      5      [0]  0.235342  0.234287   0.204871\n",
       "includes_functional_group            6      [0]  0.250199  0.249562   0.220505\n",
       "includes_paper_artifacts             7      [0]  0.132717  0.132419   0.122461\n",
       "includes_no_terms                    8      [0]  0.131980  0.130467   0.113957\n",
       "structure_adjacent_mentions          9      [0]  0.092906  0.092906   0.082430\n",
       "structure_sep_or                    10      [0]  0.015873  0.015733   0.013602\n",
       "structure_sep_comma                 11      [0]  0.048116  0.047897   0.041385\n",
       "structure_sep_via                   12      [0]  0.003664  0.003605   0.003226\n",
       "structure_sep_sym                   13      [0]  0.174938  0.173664   0.148929\n",
       "structure_sep_adverb                14      [0]  0.000697  0.000697   0.000637\n",
       "includes_oxidation_words            15      [1]  0.086673  0.086653   0.084522\n",
       "structure_sep_verb                  16      [1]  0.009858  0.009838   0.009619\n",
       "structure_sep_conversion_words      17      [1]  0.024636  0.024636   0.022783\n",
       "includes_combustion_words           18      [1]  0.001215  0.001215   0.001215\n",
       "includes_neutralization_words       19      [1]  0.045547  0.045547   0.045547\n",
       "includes_combination_words          20      [1]  0.024855  0.024835   0.024496\n",
       "includes_decomposition_words        21      [1]  0.001135  0.001115   0.001115\n",
       "includes_replacement_words          22      [1]  0.008763  0.008703   0.008604\n",
       "includes_reaction_words             23      [1]  0.750737  0.745081   0.727256\n",
       "includes_reaction_component_words   24      [1]  0.146459  0.146061   0.142038\n",
       "includes_comparison_words           25      [0]  0.236975  0.235601   0.204055\n",
       "includes_one_chem                   26      [0]  0.109297  0.108042   0.090337\n",
       "includes_react                      27      [1]  0.098502  0.098502   0.095256\n",
       "includes_measure_words              28      [0]  0.576257  0.571298   0.498486\n",
       "includes_experiment_words           29      [0]  0.365550  0.362981   0.318071\n",
       "includes_concentration              30      [1]  0.052358  0.052358   0.050944\n",
       "structure_sep_and                   31      [0]  0.086513  0.086035   0.076097\n",
       "structure_next_to_conversion_words  32      [1]  0.206883  0.206026   0.197443\n",
       "structure_conversion_by             33      [1]  0.008165  0.008165   0.007190\n",
       "structure_conversion_of             34      [1]  0.114933  0.114933   0.110511\n",
       "includes_react_sym                  35      [1]  0.000080  0.000080   0.000080\n",
       "includes_interconvert_sym           36       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_1           37      [1]  0.447045  0.446965   0.432347\n",
       "structure_jtsui_pattern_2           38      [1]  0.152215  0.152215   0.146539\n",
       "structure_jtsui_pattern_3           39       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_4           40       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_5           41       []  0.000000  0.000000   0.000000"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from snorkel.labeling import LFAnalysis\n",
    "\n",
    "LFAnalysis(L=L_train, lfs=lfs).lf_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9933681191746992"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LFAnalysis(L_train).label_coverage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-5.228013223930534"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gets the average amount of labels (goal is to get this to 15)\n",
    "sum = 0\n",
    "total = 0\n",
    "for item in L_train:\n",
    "    sum += 31 - np.count_nonzero(item == -1)\n",
    "    total += 1\n",
    "sum/total\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "from snorkel.labeling.model import MajorityLabelVoter\n",
    "from snorkel.labeling.model import LabelModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 1414/1414 [00:12<00:00, 109.16it/s]\n"
     ]
    }
   ],
   "source": [
    "# Define the set of labeling functions (LFs)\n",
    "# currently excluding amino_acid and followed_ase and followed_by_noun\n",
    "\n",
    "# removing physical_words increases recall but causes large drop in precision\n",
    "# sep_conversion_word and sep_verb removal increase precision to 0.71 with recall at 0.38\n",
    "# Apply the LFs to the unlabeled training data\n",
    "df_test = sentence_chem_df\n",
    "applier_test = PandasLFApplier(lfs=lfs)\n",
    "L_test = applier_test.apply(df=df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>includes_solution_words</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.012023</td>\n",
       "      <td>0.012023</td>\n",
       "      <td>0.011315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_mixture_words</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.001414</td>\n",
       "      <td>0.001414</td>\n",
       "      <td>0.001414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_physical_words</th>\n",
       "      <td>2</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.304809</td>\n",
       "      <td>0.304809</td>\n",
       "      <td>0.225601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_genetic_words</th>\n",
       "      <td>3</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.205799</td>\n",
       "      <td>0.205799</td>\n",
       "      <td>0.162659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_structural_words</th>\n",
       "      <td>4</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.194484</td>\n",
       "      <td>0.193777</td>\n",
       "      <td>0.152758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_general_chemical_words</th>\n",
       "      <td>5</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.114569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_functional_group</th>\n",
       "      <td>6</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.173975</td>\n",
       "      <td>0.173975</td>\n",
       "      <td>0.144272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_paper_artifacts</th>\n",
       "      <td>7</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.041726</td>\n",
       "      <td>0.041726</td>\n",
       "      <td>0.029703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_no_terms</th>\n",
       "      <td>8</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.192362</td>\n",
       "      <td>0.191655</td>\n",
       "      <td>0.145686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_adjacent_mentions</th>\n",
       "      <td>9</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.396040</td>\n",
       "      <td>0.396040</td>\n",
       "      <td>0.275106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_or</th>\n",
       "      <td>10</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.027581</td>\n",
       "      <td>0.027581</td>\n",
       "      <td>0.020509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_comma</th>\n",
       "      <td>11</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.200849</td>\n",
       "      <td>0.200849</td>\n",
       "      <td>0.133663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_via</th>\n",
       "      <td>12</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.167610</td>\n",
       "      <td>0.167610</td>\n",
       "      <td>0.121641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_sym</th>\n",
       "      <td>13</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.400990</td>\n",
       "      <td>0.398161</td>\n",
       "      <td>0.275813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_adverb</th>\n",
       "      <td>14</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_oxidation_words</th>\n",
       "      <td>15</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.090523</td>\n",
       "      <td>0.090523</td>\n",
       "      <td>0.089109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_verb</th>\n",
       "      <td>16</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.006365</td>\n",
       "      <td>0.006365</td>\n",
       "      <td>0.006365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_conversion_words</th>\n",
       "      <td>17</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.096888</td>\n",
       "      <td>0.096888</td>\n",
       "      <td>0.095474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combustion_words</th>\n",
       "      <td>18</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_neutralization_words</th>\n",
       "      <td>19</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.003536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_combination_words</th>\n",
       "      <td>20</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.021216</td>\n",
       "      <td>0.021216</td>\n",
       "      <td>0.021216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_decomposition_words</th>\n",
       "      <td>21</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_replacement_words</th>\n",
       "      <td>22</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.002829</td>\n",
       "      <td>0.002829</td>\n",
       "      <td>0.002829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_words</th>\n",
       "      <td>23</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.591231</td>\n",
       "      <td>0.591231</td>\n",
       "      <td>0.585573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_reaction_component_words</th>\n",
       "      <td>24</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.076379</td>\n",
       "      <td>0.076379</td>\n",
       "      <td>0.074965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_comparison_words</th>\n",
       "      <td>25</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.262376</td>\n",
       "      <td>0.260962</td>\n",
       "      <td>0.193777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_one_chem</th>\n",
       "      <td>26</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.350778</td>\n",
       "      <td>0.345120</td>\n",
       "      <td>0.251768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_react</th>\n",
       "      <td>27</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.026874</td>\n",
       "      <td>0.026874</td>\n",
       "      <td>0.026167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_measure_words</th>\n",
       "      <td>28</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.393211</td>\n",
       "      <td>0.392504</td>\n",
       "      <td>0.302687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_experiment_words</th>\n",
       "      <td>29</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.484441</td>\n",
       "      <td>0.480905</td>\n",
       "      <td>0.357850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_concentration</th>\n",
       "      <td>30</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.031117</td>\n",
       "      <td>0.031117</td>\n",
       "      <td>0.031117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_sep_and</th>\n",
       "      <td>31</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.227016</td>\n",
       "      <td>0.224894</td>\n",
       "      <td>0.164074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_next_to_conversion_words</th>\n",
       "      <td>32</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.239745</td>\n",
       "      <td>0.239745</td>\n",
       "      <td>0.237624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_conversion_by</th>\n",
       "      <td>33</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.002122</td>\n",
       "      <td>0.002122</td>\n",
       "      <td>0.002122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_conversion_of</th>\n",
       "      <td>34</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.108204</td>\n",
       "      <td>0.108204</td>\n",
       "      <td>0.106789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_react_sym</th>\n",
       "      <td>35</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.000707</td>\n",
       "      <td>0.000707</td>\n",
       "      <td>0.000707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>includes_interconvert_sym</th>\n",
       "      <td>36</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_1</th>\n",
       "      <td>37</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.301980</td>\n",
       "      <td>0.301980</td>\n",
       "      <td>0.296322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_2</th>\n",
       "      <td>38</td>\n",
       "      <td>[1]</td>\n",
       "      <td>0.096888</td>\n",
       "      <td>0.096888</td>\n",
       "      <td>0.094767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_3</th>\n",
       "      <td>39</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_4</th>\n",
       "      <td>40</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>structure_jtsui_pattern_5</th>\n",
       "      <td>41</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     j Polarity  Coverage  Overlaps  Conflicts\n",
       "includes_solution_words              0      [0]  0.012023  0.012023   0.011315\n",
       "includes_mixture_words               1      [0]  0.001414  0.001414   0.001414\n",
       "includes_physical_words              2      [0]  0.304809  0.304809   0.225601\n",
       "includes_genetic_words               3      [0]  0.205799  0.205799   0.162659\n",
       "includes_structural_words            4      [0]  0.194484  0.193777   0.152758\n",
       "includes_general_chemical_words      5      [0]  0.142857  0.142857   0.114569\n",
       "includes_functional_group            6      [0]  0.173975  0.173975   0.144272\n",
       "includes_paper_artifacts             7      [0]  0.041726  0.041726   0.029703\n",
       "includes_no_terms                    8      [0]  0.192362  0.191655   0.145686\n",
       "structure_adjacent_mentions          9      [0]  0.396040  0.396040   0.275106\n",
       "structure_sep_or                    10      [0]  0.027581  0.027581   0.020509\n",
       "structure_sep_comma                 11      [0]  0.200849  0.200849   0.133663\n",
       "structure_sep_via                   12      [0]  0.167610  0.167610   0.121641\n",
       "structure_sep_sym                   13      [0]  0.400990  0.398161   0.275813\n",
       "structure_sep_adverb                14       []  0.000000  0.000000   0.000000\n",
       "includes_oxidation_words            15      [1]  0.090523  0.090523   0.089109\n",
       "structure_sep_verb                  16      [1]  0.006365  0.006365   0.006365\n",
       "structure_sep_conversion_words      17      [1]  0.096888  0.096888   0.095474\n",
       "includes_combustion_words           18       []  0.000000  0.000000   0.000000\n",
       "includes_neutralization_words       19      [1]  0.003536  0.003536   0.003536\n",
       "includes_combination_words          20      [1]  0.021216  0.021216   0.021216\n",
       "includes_decomposition_words        21       []  0.000000  0.000000   0.000000\n",
       "includes_replacement_words          22      [1]  0.002829  0.002829   0.002829\n",
       "includes_reaction_words             23      [1]  0.591231  0.591231   0.585573\n",
       "includes_reaction_component_words   24      [1]  0.076379  0.076379   0.074965\n",
       "includes_comparison_words           25      [0]  0.262376  0.260962   0.193777\n",
       "includes_one_chem                   26      [0]  0.350778  0.345120   0.251768\n",
       "includes_react                      27      [1]  0.026874  0.026874   0.026167\n",
       "includes_measure_words              28      [0]  0.393211  0.392504   0.302687\n",
       "includes_experiment_words           29      [0]  0.484441  0.480905   0.357850\n",
       "includes_concentration              30      [1]  0.031117  0.031117   0.031117\n",
       "structure_sep_and                   31      [0]  0.227016  0.224894   0.164074\n",
       "structure_next_to_conversion_words  32      [1]  0.239745  0.239745   0.237624\n",
       "structure_conversion_by             33      [1]  0.002122  0.002122   0.002122\n",
       "structure_conversion_of             34      [1]  0.108204  0.108204   0.106789\n",
       "includes_react_sym                  35      [1]  0.000707  0.000707   0.000707\n",
       "includes_interconvert_sym           36       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_1           37      [1]  0.301980  0.301980   0.296322\n",
       "structure_jtsui_pattern_2           38      [1]  0.096888  0.096888   0.094767\n",
       "structure_jtsui_pattern_3           39       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_4           40       []  0.000000  0.000000   0.000000\n",
       "structure_jtsui_pattern_5           41       []  0.000000  0.000000   0.000000"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LFAnalysis(L=L_test, lfs=lfs).lf_summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9985855728429985"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LFAnalysis(L_test).label_coverage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/338320555.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_test[\"label_voter\"] = majority_model.predict(L=L_test)\n"
     ]
    }
   ],
   "source": [
    "majority_model = MajorityLabelVoter()\n",
    "df_test[\"label_voter\"] = majority_model.predict(L=L_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\cassi\\AppData\\Local\\Temp/ipykernel_17872/488278135.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_test[\"label_model\"] = label_model.predict(L=L_test)\n"
     ]
    }
   ],
   "source": [
    "from snorkel.labeling.model import LabelModel\n",
    "\n",
    "label_model = LabelModel(cardinality=2, verbose=True)\n",
    "label_model.fit(L_train=L_train, n_epochs=500, log_freq=100, seed=123, class_balance=[0.95, 0.05])\n",
    "df_test[\"label_model\"] = label_model.predict(L=L_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_second(probs):\n",
    "    return (probs[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_train[\"label_probs\"] = np.apply_along_axis(grab_second, 1, majority_model.predict_proba(L=L_train))\n",
    "# df_train[\"label_probs\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_test.to_csv(r'../labeled.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1354\n",
       "1      60\n",
       "Name: truth, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test[\"truth\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>chemicals</th>\n",
       "      <th>truth</th>\n",
       "      <th>substrates</th>\n",
       "      <th>products</th>\n",
       "      <th>text</th>\n",
       "      <th>label_voter</th>\n",
       "      <th>label_model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>the enzyme cyclo-oxygenase catalyses the oxyge...</td>\n",
       "      <td>[prostaglandins]</td>\n",
       "      <td>1</td>\n",
       "      <td>arachidonic, acid</td>\n",
       "      <td>prostaglandins</td>\n",
       "      <td>[the enzyme cyclo-oxygenase catalyses the oxyg...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>recently two forms of cyclo-oxygenase have bee...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[recently two forms of cyclo-oxygenase have be...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>constitutive and inducible forms of human cycl...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[constitutive and inducible forms of human cyc...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hcox-1 had a specific activity of 188 mumol of...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[hcox-1 had a specific activity of 188 mumol o...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>of 1500 nmol of o2/nmol of enzyme, whereas hco...</td>\n",
       "      <td>[arachidonate]</td>\n",
       "      <td>0</td>\n",
       "      <td>arachidonate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[of 1500 nmol of o2/nmol of enzyme, whereas hc...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>water uptake of the polymer was only 28 and 02...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[water uptake of the polymer was only 28 and 0...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1410</th>\n",
       "      <td>degradation of less hydrophilic pec41 with hig...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[degradation of less hydrophilic pec41 with hi...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1411</th>\n",
       "      <td>by this mechanism, ce-responsive drug in vitro...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[by this mechanism, ce-responsive drug in vitr...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1412</th>\n",
       "      <td>as expected, less bovine serum albumin bsa was...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[as expected, less bovine serum albumin bsa wa...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1413</th>\n",
       "      <td>in conclusion, this work enabled the in vitro ...</td>\n",
       "      <td>[pec]</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[in conclusion, this work enabled the in vitro...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1414 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               sentence         chemicals  \\\n",
       "0     the enzyme cyclo-oxygenase catalyses the oxyge...  [prostaglandins]   \n",
       "1     recently two forms of cyclo-oxygenase have bee...                []   \n",
       "2     constitutive and inducible forms of human cycl...                []   \n",
       "3     hcox-1 had a specific activity of 188 mumol of...    [arachidonate]   \n",
       "4     of 1500 nmol of o2/nmol of enzyme, whereas hco...    [arachidonate]   \n",
       "...                                                 ...               ...   \n",
       "1409  water uptake of the polymer was only 28 and 02...             [pec]   \n",
       "1410  degradation of less hydrophilic pec41 with hig...                []   \n",
       "1411  by this mechanism, ce-responsive drug in vitro...             [pec]   \n",
       "1412  as expected, less bovine serum albumin bsa was...                []   \n",
       "1413  in conclusion, this work enabled the in vitro ...             [pec]   \n",
       "\n",
       "      truth         substrates        products  \\\n",
       "0         1  arachidonic, acid  prostaglandins   \n",
       "1         0                NaN             NaN   \n",
       "2         0                NaN             NaN   \n",
       "3         0       arachidonate             NaN   \n",
       "4         0       arachidonate             NaN   \n",
       "...     ...                ...             ...   \n",
       "1409      0                NaN             NaN   \n",
       "1410      0                NaN             NaN   \n",
       "1411      0                NaN             NaN   \n",
       "1412      0                NaN             NaN   \n",
       "1413      0                NaN             NaN   \n",
       "\n",
       "                                                   text  label_voter  \\\n",
       "0     [the enzyme cyclo-oxygenase catalyses the oxyg...            1   \n",
       "1     [recently two forms of cyclo-oxygenase have be...            0   \n",
       "2     [constitutive and inducible forms of human cyc...            0   \n",
       "3     [hcox-1 had a specific activity of 188 mumol o...            0   \n",
       "4     [of 1500 nmol of o2/nmol of enzyme, whereas hc...            0   \n",
       "...                                                 ...          ...   \n",
       "1409  [water uptake of the polymer was only 28 and 0...            0   \n",
       "1410  [degradation of less hydrophilic pec41 with hi...            0   \n",
       "1411  [by this mechanism, ce-responsive drug in vitr...            1   \n",
       "1412  [as expected, less bovine serum albumin bsa wa...            0   \n",
       "1413  [in conclusion, this work enabled the in vitro...            0   \n",
       "\n",
       "      label_model  \n",
       "0               0  \n",
       "1               0  \n",
       "2               0  \n",
       "3               0  \n",
       "4               0  \n",
       "...           ...  \n",
       "1409            0  \n",
       "1410            0  \n",
       "1411            0  \n",
       "1412            0  \n",
       "1413            0  \n",
       "\n",
       "[1414 rows x 8 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "only_truth_df = df_test.dropna(subset=['truth'])\n",
    "only_truth_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sklearn as sk\n",
    "from sklearn.metrics import recall_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import balanced_accuracy_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "f1: 0.38317757009345793\n",
      "recall: 0.7454545454545455\n",
      "precision: 0.2578616352201258\n",
      "accuracy: 0.8980694980694981\n",
      "balanced accuracy: 0.8251466275659824\n",
      "matrix: (1122, 118, 14, 41)\n"
     ]
    }
   ],
   "source": [
    "no_abstain_df = only_truth_df[only_truth_df[\"label_voter\"] != -1]\n",
    "print(\"f1:\", sk.metrics.f1_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]))\n",
    "print(\"recall:\", recall_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]))\n",
    "print(\"precision:\", precision_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]))\n",
    "print(\"accuracy:\", accuracy_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]))\n",
    "print(\"balanced accuracy:\", balanced_accuracy_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]))\n",
    "tn, fp, fn, tp = confusion_matrix(no_abstain_df[\"truth\"], no_abstain_df[\"label_voter\"]).ravel()\n",
    "print(\"matrix:\", (tn, fp, fn, tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "f1: 0.4833333333333333\n",
    "recall: 0.5576923076923077\n",
    "precision: 0.4264705882352941\n",
    "accuracy: 0.9533483822422875\n",
    "balanced accuracy: 0.763575989398229\n",
    "matrix: (1238, 39, 23, 29)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "f1: 0.21951219512195122\n",
      "recall: 0.15\n",
      "precision: 0.4090909090909091\n",
      "accuracy: 0.9547383309759547\n",
      "balanced accuracy: 0.5701994091580502\n",
      "matrix: (1341, 13, 51, 9)\n"
     ]
    }
   ],
   "source": [
    "no_abstain_df = only_truth_df[only_truth_df[\"label_model\"] != -1]\n",
    "print(\"f1:\", sk.metrics.f1_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]))\n",
    "print(\"recall:\", recall_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]))\n",
    "print(\"precision:\", precision_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]))\n",
    "print(\"accuracy:\", accuracy_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]))\n",
    "print(\"balanced accuracy:\", balanced_accuracy_score(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]))\n",
    "tn, fp, fn, tp = confusion_matrix(no_abstain_df[\"truth\"], no_abstain_df[\"label_model\"]).ravel()\n",
    "print(\"matrix:\", (tn, fp, fn, tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "f1: 0.14925373134328357\n",
    "recall: 0.08333333333333333\n",
    "precision: 0.7142857142857143\n",
    "accuracy: 0.9596888260254597\n",
    "balanced accuracy: 0.5409281142294436\n",
    "matrix: (1352, 2, 55, 5)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'label_voter'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python38\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3360\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3361\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python38\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python38\\site-packages\\pandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'label_voter'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_18152/277158610.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mdf_train\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mdf_train\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"label_voter\"\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"truth\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalue_counts\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python38\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3456\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3457\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3458\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3459\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3460\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python38\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3361\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3362\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3363\u001b[1;33m                 \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3364\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3365\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mis_scalar\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhasnans\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'label_voter'"
     ]
    }
   ],
   "source": [
    "df_train[df_train[\"label_voter\"] != -1][\"truth\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1285\n",
       "1      45\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train[df_train[\"label_voter\"] != -1][\"label\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    323\n",
       "1      7\n",
       "Name: label, dtype: int64"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_df = df_train[df_train[\"label_voter\"] == df_train[\"truth\"]]\n",
    "matched_df[\"label\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e1fffad673803e23a98281b7492a34c559235864658941242758d49259d94b3b"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit (windows store)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
